// @sumup/foundry/presets/index.js
"use strict";
/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.presetChoices = exports.presets = void 0;
const chalk_1 = __importDefault(require("chalk"));
const shared_1 = require("../types/shared");
const lint = {
    name: formatName('Lint', 'Check code for syntax errors and format it automatically'),
    value: shared_1.Preset.LINT,
    short: 'Lint',
    tools: [shared_1.Tool.ESLINT, shared_1.Tool.PRETTIER, shared_1.Tool.HUSKY, shared_1.Tool.LINT_STAGED],
    prompts: [
        shared_1.Prompt.LANGUAGE,
        shared_1.Prompt.ENVIRONMENTS,
        shared_1.Prompt.FRAMEWORKS,
        shared_1.Prompt.OPEN_SOURCE,
    ],
};
const templates = {
    name: formatName('Templates', 'Generate boilerplate code e.g. for React components'),
    value: shared_1.Preset.TEMPLATES,
    short: 'Templates',
    tools: [shared_1.Tool.PLOP],
    prompts: [shared_1.Prompt.LANGUAGE],
};
const release = {
    name: formatName('Release', 'Automatically generate release notes and (optionally) publish to NPM'),
    value: shared_1.Preset.RELEASE,
    short: 'Release',
    tools: [shared_1.Tool.SEMANTIC_RELEASE],
    prompts: [shared_1.Prompt.PUBLISH],
};
const ci = {
    name: formatName('Continous Integration', 'Validate the code on every push using the configured presets'),
    value: shared_1.Preset.CI,
    short: 'CI',
    tools: [shared_1.Tool.CI],
    prompts: [shared_1.Prompt.CI],
};
function formatName(name, description) {
    return [`${chalk_1.default.bold(name)}:`, description].join(' ');
}
exports.presets = { lint, templates, release, ci };
exports.presetChoices = [lint, templates, release, ci];
