// @sumup/foundry/lib/logger.js
"use strict";
/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.empty = exports.debug = exports.info = exports.log = exports.warn = exports.error = void 0;
/* eslint-disable no-console */
const chalk_1 = __importDefault(require("chalk"));
const IS_DEBUG = process.argv.includes('--debug') || process.env.NODE_ENV === 'DEBUG';
const getMessage = (arg) => {
    const message = Array.isArray(arg) ? arg.join('\n') : arg;
    return message;
};
exports.error = (arg) => {
    const msg = getMessage(arg);
    console.error(`ðŸš¨ ${chalk_1.default.red(msg)}`);
};
exports.warn = (arg) => {
    const msg = getMessage(arg);
    console.warn(`âš ï¸ ${chalk_1.default.yellow(msg)}`);
};
exports.log = (arg) => {
    const msg = getMessage(arg);
    console.log(msg);
};
exports.info = (arg) => {
    const msg = getMessage(arg);
    console.log(`â„¹ï¸ ${msg}`);
};
exports.debug = (arg) => {
    if (!IS_DEBUG) {
        return;
    }
    const msg = getMessage(arg);
    console.debug(`ðŸ› ï¸ ${chalk_1.default.cyan(msg)}`);
};
exports.empty = () => {
    console.log('');
};
