// @sumup/foundry/configs/plop/index.js
"use strict";
/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.scripts = exports.files = void 0;
const fp_1 = require("lodash/fp");
exports.files = (options) => [
    {
        name: 'plopfile.js',
        content: `
    module.exports = require('@sumup/foundry/plop')(${JSON.stringify(fp_1.pick(['language'], options))})`,
    },
];
exports.scripts = () => [
    {
        name: 'create:component',
        command: 'foundry run plop component',
        description: 'generate the files for a React component',
    },
];
