// @sumup/foundry/cli/index.js
#!/usr/bin/env node
"use strict";
/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const yargs_1 = __importDefault(require("yargs"));
const shared_1 = require("../types/shared");
const choices_1 = require("../lib/choices");
const run_1 = require("./run");
const init_1 = require("./init");
// eslint-disable-next-line @typescript-eslint/no-unused-expressions
yargs_1.default
    .command('init', "Initialize Foundry's tools in your project", {
    presets: {
        alias: 'p',
        desc: 'A preset configures a group of tools that solve a common problem',
        choices: choices_1.enumToChoices(shared_1.Preset),
        type: 'array',
    },
    language: {
        alias: 'l',
        desc: 'The programming language the project uses',
        choices: choices_1.enumToChoices(shared_1.Language),
    },
    environments: {
        alias: 'e',
        desc: 'The environment(s) that the code runs in',
        choices: choices_1.enumToChoices(shared_1.Environment),
        type: 'array',
    },
    frameworks: {
        alias: 'f',
        desc: 'The frameworks the project uses',
        choices: choices_1.enumToChoices(shared_1.Framework),
        type: 'array',
    },
    ci: {
        desc: 'The CI platform the project uses',
        choices: choices_1.enumToChoices(shared_1.CI),
    },
    openSource: {
        alias: 'o',
        desc: 'Whether the project is open source',
        type: 'boolean',
    },
    publish: {
        desc: 'Whether to publish to NPM',
        type: 'boolean',
    },
    configDir: {
        alias: 'c',
        desc: 'The directory to write configs to',
        type: 'string',
        default: '.',
    },
    overwrite: {
        desc: 'Whether to overwrite existing config files',
        type: 'boolean',
        default: false,
    },
}, execute('init'))
    .command('run <tool> [...tool options]', 'Run any of the bundled tools.', execute('run'))
    .showHelpOnFail(true)
    .demandCommand(1, '')
    .help()
    .version().argv;
function execute(command) {
    const commands = { run: run_1.run, init: init_1.init };
    const commandFn = commands[command];
    return (args) => {
        commandFn(args);
    };
}
